File: Algorithms/p101_Symmetric_Tree_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return 1;
        stack<TreeNode*,vector<TreeNode*>> stack1;
        stack<TreeNode*,vector<TreeNode*>> stack2;
        TreeNode* cur1 = root;
        TreeNode* cur2 = root;
        while(1){
            if(cur1&&cur2){
                if(cur1->val!=cur2->val) return 0;
                stack1.push(cur1->right);
                stack2.push(cur2->left);
                cur1 = cur1->left;
                cur2 = cur2->right;
            }else if(!cur1&&!cur2){
                if(stack1.empty()&&stack2.empty()) break;
                if(stack1.empty()||stack2.empty()) return 0;
                cur1 = stack1.top();
                cur2 = stack2.top();
                stack1.pop();
                stack2.pop();
            }else{
                return 0;
            }
        }
        return 1;
    }
};

File: Algorithms/p101_Symmetric_Tree/p101_Symmetric_Tree_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        stack1 = []
        stack2 = []
        cur1 = root
        cur2 = root
        while True:
            if cur1 and cur2:
                if cur1.val!=cur2.val:
                    return False
                stack1.append(cur1.right)
                stack2.append(cur2.left)
                cur1 = cur1.left
                cur2 = cur2.right
            elif not cur1 and not cur2:
                if not stack1 and not stack2:
                    break
                if not stack1 or not stack2:
                    return False
                cur1 = stack1.pop()
                cur2 = stack2.pop()
            else:
                return False
        return True
